calico_node/startup/startup.go

```diff
@@ -828,6 +834,68 @@ func ensureNodeFelixConfig(c *client.Client, host, key, def string) error {
 	}
 }
 
+// ensureClusterTypeSet ensures that the ClusterType in the Felix global
+// config has the data contained in the passed in clusterType.  This includes
+// merging what was already set in ClusterType with the passed in clusterType.
+func ensureClusterTypeSet(c *client.Client, key, clusterType string) error {
+	global := ""
+	// Do this twice for the case where we update the field but someone else
+	// overwrites our update, don't do this forever because we don't want to
+	// take too long making this update.
+	for i := 0; i < 2; i++ {
+		if val, assigned, err := c.Config().GetFelixConfig("ClusterType", global); err != nil {
+			return err
+		} else if !assigned {
+			if err := c.Config().SetFelixConfig(key, global, clusterType); err != nil {
+				return err
+			}
+		} else {
+			// Sort so any duplicates will be consecutive
+			all := append(strings.Split(val, ","), strings.Split(clusterType, ",")...)
+			sort.Strings(all)
+
+			if len(all) == 0 {
+				// No cluster type at all? nothing to do
+				return nil
+			}
+
+			// A slice to hold uniq strings
+			types := []string{all[0]}
+			x := 0
+			for j := 1; j < len(all); j++ {
+				if types[x] == all[j] {
+					continue
+				}
+
+				types = append(types, all[j])
+				x++
+			}
+
+			newClusterTypes := strings.Join(types, ",")
+
+			// If current field already is what we want then nothing to do
+			if strings.Compare(newClusterTypes, val) == 0 {
+				return nil
+			}
+
+			// Try doing a get before doing our update to try to make the window
```

Comment:
A spec question.  Who sets the cluster type, and what happens if all of the nodes of a particular cluster are deleted - we'd never remove that cluster type from the list would we?  I wonder if the cluster type and version should really have been per-node.  I'm sure this has been discussed, but raising here so I remember to ask tomorrow :-)